en: add doc for enable TLS between TiDB components (#64)
pingcap/docs-tidb-operator: en/enable-tls-between-components.md

---
title: Enable TLS between TiDB Components
summary: Learn how to enable TLS between TiDB components on Kubernetes.
category: how-to
---

# Enable TLS between TiDB Components

This document describes how to enable Transport Layer Security (TLS) between components of the TiDB cluster in Kubernetes, which is supported since TiDB Operator v1.1.

To enable TLS between TiDB components, perform the following steps:

1. Generate certificates for each component of the TiDB cluster to be created:

   - A set of server-side certificates for the PD/TiKV/TiDB/Pump/Drainer component, saved as the Kubernetes Secret objects: `<cluster-name>-<component-name>-cluster-secret`

>    - A set of shared client-side certificates for the various clients of each component, saved as the Kubernetes Secret objects:  `<cluster-name>-cluster-client-secret`.

Reviewer: ```suggestion
   - A set of shared client-side certificates for the various clients of each component, saved as the Kubernetes Secret objects: `<cluster-name>-cluster-client-secret`.
```

    mkdir -p ~/bin
    curl -s -L -o ~/bin/cfssl https://pkg.cfssl.org/R1.2/cfssl_linux-amd64
    curl -s -L -o ~/bin/cfssljson https://pkg.cfssl.org/R1.2/cfssljson_linux-amd64
    chmod +x ~/bin/{cfssl,cfssljson}
    export PATH=$PATH:~/bin
    mkdir -p ~/cfssl
    cd ~/cfssl
    cfssl print-defaults config > ca-config.json
    cfssl print-defaults csr > ca-csr.json
    ```

2. Configure the client auth (CA) option in `ca-config.json`:

    ```json
    {
        "signing": {
            "default": {
                "expiry": "8760h"
            },
            "profiles": {
                "server": {
                    "expiry": "8760h",
                    "usages": [
                        "signing",
                        "key encipherment",
                        "server auth",
                        "client auth"
                    ]
                },
                "client": {
                    "expiry": "8760h",
                    "usages": [
                        "signing",
                        "key encipherment",
                        "client auth"
                    ]
                }
            }
        }
    }
    ```

    > **Note:**
    >

>     > Add `"client auth"` in `profile`-`server`- `usages`, because this server-side certificate is also used as the client-side certificate.

Reviewer: ```suggestion
    > Add `"client auth"` in `profiles` - `server` - `usages`, because this server-side certificate is also used as the client-side certificate.
```

    kind: Certificate
    metadata:
      name: tidb-cluster-issuer-cert
      namespace: cert-manager
    spec:
      secretName: tidb-cluster-issuer-cert
      commonName: "TiDB CA"
      isCA: true
      issuerRef:
        name: tidb-selfsigned-ca-issuer
        kind: ClusterIssuer
    ---
    apiVersion: cert-manager.io/v1alpha2
    kind: ClusterIssuer
    metadata:
      name: tidb-cluster-issuer
    spec:
      ca:
        secretName: tidb-cluster-issuer-cert
    ```

    This `.yaml` file creates three objects:

    - A ClusterIssuer object of SelfSigned class, used to generate the CA certificate needed by ClusterIssuer of CA class
    - A Certificate object, whose `isCa` is set to `true`
    - A ClusterIssuer, used to issue TLS certificates between TiDB components

    Finally, execute the following command to create a ClusterIssuer:

    {{< copyable "shell-regular" >}}

    ```shell
    kubectl apply -f ~/cert-manager/tidb-cluster-issuer.yaml
    ```

3. Generate the server-side certificate.

    In `cert-manager`, the Certificate resource represents the certificate interface. This certificate is issued and updated by the ClusterIssuer created in Step 2.

>     According to [Enable TLS Authentication | TiDB Documentation](https://pingcap.com/docs/stable/how-to/secure/enable-tls-between-components/), each component needs a server-side certificate, and all components need a common client-side certificate for their clients.

Reviewer 2: ```suggestion
    According to [Enable TLS Authentication | TiDB Documentation](https://pingcap.com/docs/stable/how-to/secure/enable-tls-between-components/), each component needs a server-side certificate, and all components need a shared client-side certificate for their clients.
```

        kind: Certificate
        metadata:
          name: <cluster-name>-pd-cluster-secret
          namespace: <namespace>
        spec:
          secretName: <cluster-name>-pd-cluster-secret
          duration: 8760h # 365d
          renewBefore: 360h # 15d
          organization:
          - PingCAP
          commonName: "PD"
          usages:
            - server auth
            - client auth
          dnsNames:
          - "<cluster-name>-pd"
          - "<cluster-name>-pd.<namespace>"
          - "<cluster-name>-pd.<namespace>.svc"
          - "<cluster-name>-pd-peer"
          - "<cluster-name>-pd-peer.<namespace>"
          - "<cluster-name>-pd-peer.<namespace>.svc"
          - "*.<cluster-name>-pd-peer"
          - "*.<cluster-name>-pd-peer.<namespace>"
          - "*.<cluster-name>-pd-peer.<namespace>.svc"
          ipAddresses:
          - 127.0.0.1
          - ::1
          issuerRef:
            name: tidb-cluster-issuer
            kind: ClusterIssuer
            group: cert-manager.io
        ```

        `<cluster-name>` is the name of the cluster. Configure the items as follows:

        - Set `spec.secretName` to `<cluster-name>-pd-cluster-secret`

>         - Add `server auth` and `client auth` in `usages`

Reviewer 2: Please end these sentences with a period when it applies.

        - Add the following DNSs in `dnsNames`. You can also add other DNSs according to your needs:
            - `<cluster-name>-pd`
            - `<cluster-name>-pd.<namespace>`
            - `<cluster-name>-pd.<namespace>.svc`
            - `<cluster-name>-pd-peer`
            - `<cluster-name>-pd-peer.<namespace>`
            - `<cluster-name>-pd-peer.<namespace>.svc`
            - `*.<cluster-name>-pd-peer`
            - `*.<cluster-name>-pd-peer.<namespace>`
            - `*.<cluster-name>-pd-peer.<namespace>.svc`

>         - Add the following 2 IPs in `ipAddresses`. You can also add other IPs according to your needs:

Reviewer 2: ```suggestion
        - Add the following two IPs in `ipAddresses`. You can also add other IPs according to your needs:
```

            - `127.0.0.1`
            - `::1`

>         - Add the ClusterIssuer created above in the `issuerRef`

Reviewer 2: ```suggestion
        - Add the ClusterIssuer created above in `issuerRef`.
```

        metadata:
          name: <cluster-name>-tikv-cluster-secret
          namespace: <namespace>
        spec:
          secretName: <cluster-name>-tikv-cluster-secret
          duration: 8760h # 365d
          renewBefore: 360h # 15d
          organization:
          - PingCAP
          commonName: "TiKV"
          usages:
            - server auth
            - client auth
          dnsNames:
          - "<cluster-name>-tikv"
          - "<cluster-name>-tikv.<namespace>"
          - "<cluster-name>-tikv.<namespace>.svc"
          - "<cluster-name>-tikv-peer"
          - "<cluster-name>-tikv-peer.<namespace>"
          - "<cluster-name>-tikv-peer.<namespace>.svc"
          - "*.<cluster-name>-tikv-peer"
          - "*.<cluster-name>-tikv-peer.<namespace>"
          - "*.<cluster-name>-tikv-peer.<namespace>.svc"
          ipAddresses:
          - 127.0.0.1
          - ::1
          issuerRef:
            name: tidb-cluster-issuer
            kind: ClusterIssuer
            group: cert-manager.io
        ```

        `<cluster-name>` is the name of the cluster. Configure the items as follows:

        - Set `spec.secretName` to `<cluster-name>-tikv-cluster-secret`

>         - Add `server auth` and `client auth` in `usages`

Reviewer 2: Plz end these sentences with a period when it applies.

        - Add the following DNSs in `dnsNames`. You can also add other DNSs according to your needs:
            
            - `<cluster-name>-tikv`
            - `<cluster-name>-tikv.<namespace>`
            - `<cluster-name>-tikv.<namespace>.svc`
            - `<cluster-name>-tikv-peer`
            - `<cluster-name>-tikv-peer.<namespace>`
            - `<cluster-name>-tikv-peer.<namespace>.svc`
            - `*.<cluster-name>-tikv-peer`
            - `*.<cluster-name>-tikv-peer.<namespace>`
            - `*.<cluster-name>-tikv-peer.<namespace>.svc`

        - Add the following 2 IPs in `ipAddresses`. You can also add other IPs according to your needs:
            - `127.0.0.1`
            - `::1`

>         - Add the ClusterIssuer created above in the `issuerRef`

Reviewer 2: ```suggestion
        - Add the ClusterIssuer created above in `issuerRef`.
```

        - Add the following DNSs in `dnsNames`. You can also add other DNSs according to your needs:

            - `*.<cluster-name>-pump`
            - `*.<cluster-name>-pump.<namespace>`
            - `*.<cluster-name>-pump.<namespace>.svc`

        - Add the following 2 IPs in `ipAddresses`. You can also add other IPs according to your needs:
            - `127.0.0.1`
            - `::1`
        - Add the ClusterIssuer created above in the `issuerRef`
        - For other attributes, refer to [cert-manager API](https://cert-manager.io/docs/reference/api-docs/#cert-manager.io/v1alpha2.CertificateSpec).

        After the object is created, `cert-manager` generates a `<cluster-name>-pump-cluster-secret` Secret object to be used by the Pump component of the TiDB server.

    - Drainer

        Drainer is deployed using Helm. The `dnsNames` field varies with different configuration of the `values.yaml` file.

        If you set the `drainerName` attributes when deploying Drainer as follows:

        ```yaml
        ...
        # Change the name of the statefulset and pod
        # The default is clusterName-ReleaseName-drainer

>         # Do not change the name of an existing running drainer: this is unsupported.

Reviewer 2: Please update this code comment according to the above comment for L303-L305.

        - For other attributes, refer to [cert-manager API](https://cert-manager.io/docs/reference/api-docs/#cert-manager.io/v1alpha2.CertificateSpec).

        After the object is created, `cert-manager` generates a `<cluster-name>-drainer-cluster-secret` Secret object to be used by the Drainer component of the TiDB server.

4. Generate the client-side certificate for components of the TiDB cluster.

    ```yaml
    apiVersion: cert-manager.io/v1alpha2
    kind: Certificate
    metadata:
        name: <cluster-name>-cluster-client-secret
        namespace: <namespace>
    spec:
        secretName: <cluster-name>-cluster-client-secret
        duration: 8760h # 365d
        renewBefore: 360h # 15d
        organization:
        - PingCAP
        commonName: "TiDB Components TLS Client"
        usages:
        - client auth
        issuerRef:
        name: tidb-cluster-issuer
        kind: ClusterIssuer
        group: cert-manager.io
    ```

    `<cluster-name>` is the name of the cluster. Configure the items as follows:

    - Set `spec.secretName` to `<cluster-name>-cluster-client-secret`
    - Add `client auth` in `usages`

>     - Do not configure `dnsNames` and `ipAddresses`

Reviewer 2: ```suggestion
    - You can leave `dnsNames` and `ipAddresses` empty.
```

    - Add the ClusterIssuer created above in the `issuerRef`
    - For other attributes, refer to [cert-manager API](https://cert-manager.io/docs/reference/api-docs/#cert-manager.io/v1alpha2.CertificateSpec).

    After the object is created, `cert-manager` generates a `<cluster-name>-cluster-client-secret` Secret object to be used by the clients of the TiDB components. 
    

>     To obtain the client certificate, run the following command:

Reviewer 2: Please check with the author whether there are more than one command.
```suggestion
    To obtain the client certificate, run the following commands:
```

    {{< copyable "shell-regular" >}}

    ``` shell
    mkdir -p ~/<cluster-name>-cluster-client-tls
    cd ~/<cluster-name>-cluster-client-tls
    kubectl get secret -n <namespace> <cluster-name>-cluster-client-secret  -ojsonpath='{.data.tls\.crt}' | base64 --decode > tls.crt
    kubectl get secret -n <namespace> <cluster-name>-cluster-client-secret  -ojsonpath='{.data.tls\.key}' | base64 --decode > tls.key
    kubectl get secret -n <namespace> <cluster-name>-cluster-client-secret  -ojsonpath='{.data.ca\.crt}' | base64 --decode > ca.crt
    ```

## Step 2: Deploy the TiDB cluster

> In this step, you need to finish the following operation:

Reviewer 2: ```suggestion
In this step, you need to perform the following operations:
```

       requests:
         storage: "1Gi"
       config: {}
     tikv:
       baseImage: pingcap/tikv
       replicas: 1
       requests:
         storage: "1Gi"
       config: {}
     tidb:
       baseImage: pingcap/tidb
       replicas: 1
       service:
         type: ClusterIP
       config: {}
     pump:
       baseImage: pingcap/tidb-binlog
       replicas: 1
       requests:
         storage: "1Gi"
       config: {}
    ---
    apiVersion: pingcap.com/v1alpha1
    kind: TidbMonitor
    metadata:
     name: <cluser-name>-monitor
     namespace: <namespace>
    spec:
     clusters:
     - name: <cluster-name>
     prometheus:
       baseImage: prom/prometheus
       version: v2.11.1
     grafana:
       baseImage: grafana/grafana
       version: 6.0.1
     initializer:
       baseImage: pingcap/tidb-monitor-initializer
       version: v3.0.8
     reloader:
       baseImage: pingcap/tidb-monitor-reloader
       version: v1.0.1
     imagePullPolicy: IfNotPresent
    ```

    Execute `kubectl apply -f tidb-cluster.yaml` to create a TiDB cluster.

    This operation also includes deploying a monitoring system and the Pump component.

2. Create a Drainer component and enable TLS:

>     - Method 1: Set `drainerName` when you create Drainer

Reviewer 2: period

        kubectl apply -f backup.yaml
        ```

    - Create the `restore.yaml` file and enable TLS:

         ``` yaml
        apiVersion: pingcap.com/v1alpha1
        kind: Restore
        metadata:
          name: <cluster-name>-restore
          namespace: <namespace>
        spec:
          backupType: full
          br:
            cluster: <cluster-name>
            clusterNamespace: <namespace>
            sendCredToTikv: true
            tlsCluster:
              enabled: true
          to:
            host: <host>
            secretName: <tidb-secret>
            port: 4000
            user: root
          s3:
            provider: aws
            region: <my-region>
            secretName: <s3-secret>
            bucket: <my-bucket>
            prefix: <my-folder>
        ```

        Deploy Restore:

        {{< copyable "shell-regular" >}}

        ``` shell
        kubectl apply -f restore.yaml
        ```

## Step 3: Configure `pd-ctl` and connect to the cluster

1. Download `pd-ctl`:

    Refer to [Download TiDB installation package](https://pingcap.com/docs/stable/reference/tools/pd-control/#download-tidb-installation-package).

2. Connect to the cluster:

>     First, download the client-side certificate, which is the client certificate you created in Step 1. You can directly use it, or obtain it from the `<cluster-name>-cluster-client-secret` Kubernetes Secret object created before.

Reviewer 2: ```suggestion
    First, download the client-side certificate, which is the client certificate you have created in Step 1. You can directly use it, or obtain it from the `<cluster-name>-cluster-client-secret` Kubernetes Secret object created before.
```
